src/main/java/com/atlassian/bitbucket/jenkins/internal/provider/JenkinsProvider.java

```diff
@@ -15,4 +15,12 @@
      */
     @Nonnull
     Jenkins get();
+
+    /**
+     * Returns the Jenkins root URL. This method should never be invoked from inside the extensions as the underlying
+     * implementation iterates over exisiting extensions and thus creating circular dependency.
+     *
+     * @return Jenkins root URL.
+     */
+    String getJenkinsUrl();
```

Comment:
Do we really need this method? I'm concerned that we're on a path to expose more and more of the methods on the `Jenkins` class in our own interface here.

Reply:
I think that was the idea of this interface, to help in Unit testing. Jenkins in its implementation does all kind of stuff. Classes like RetryingWebhookHandler need Jenkins URL and having only Jenkins would mean mocking a Jenkins and then overriding getRootUrl.

Reply:
But `Jenkins` is not a final class, so you can just go 
```java
Jenkins j = mock(Jenkins.class);
when(j.getRootUrl()).return("http://localhost");
```
Which is less code and prevents our testing fixtures from creeping into production code.

Reply:
I don't feel too strongly about it keeping or removing it. I have removed it for now. We can always bring it back if there are problems.